pkg/agent/plugin/nodeattestor/k8s/k8scert.go

```diff
@@ -0,0 +1,173 @@
+	}
+	return key, nil
+}
+
+func genCSR(key crypto.PrivateKey, name string) ([]byte, error) {
+	csrTemplate := x509.CertificateRequest{
+		Subject: pkix.Name{
+			// This exact format is required for the CSR to be auto-approved by kube-controller
+			// It must also NOT include any DNSNames or IPAddresses
+			Organization: []string{"system:nodes"},
+			CommonName:   "system:node:" + name,
+		},
+		SignatureAlgorithm: x509.ECDSAWithSHA256,
+	}
+	return x509.CreateCertificateRequest(rand.Reader, &csrTemplate, key)
+}
+
+func getCSRObject(csrBytes []byte, name string) *certificates.CertificateSigningRequest {
+	pemBlock := pem.Block{
+		Type:  csrPemBlockType,
+		Bytes: csrBytes,
+	}
+	pemBytes := pem.EncodeToMemory(&pemBlock)
+
+	ret := &certificates.CertificateSigningRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: certificates.CertificateSigningRequestSpec{
+			Groups:  []string{"system:authenticated"},
+			Request: pemBytes,
+			Usages:  []certificates.KeyUsage{"digital signature", "key encipherment", "client auth"},
+		},
+	}
+	return ret
+}
+
+func fetchK8sCert(kubeConfigFilePath, clientCertFilePath, clientKeyFilePath, caCertFilePath string) (*tls.Certificate, error) {
+	kubeClient, err := getKubeClient(kubeConfigFilePath, clientCertFilePath, clientKeyFilePath, caCertFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("Error creating Kubernetes client: %v", err)
+	}
+
+	key, err := genPrivateKey()
+	if err != nil {
+		return nil, fmt.Errorf("Error creating private key: %v", err)
+	}
+
+	agentName, err := os.Hostname()
+	if err != nil {
+		agentName = "unknown"
+	}
+
+	csr, err := genCSR(key, agentName)
+	if err != nil {
+		return nil, fmt.Errorf("Error creating certificate signing request: %v", err)
+	}
+
+	csrObjectName := fmt.Sprintf("%s-%d", agentName, time.Now().Unix())
+	csrObject := getCSRObject(csr, csrObjectName)
+
+	// Create CSR object in ApiServer
+	csrObject, err = kubeClient.CertificatesV1beta1().CertificateSigningRequests().Create(csrObject)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create the certificate signing request: %s", err)
+	}
+
+	// Watch for approval
+	var cert []byte
+	watcher, err := kubeClient.CertificatesV1beta1().CertificateSigningRequests().Watch(metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("Error watching CSRs on ApiServer: %v", err)
+	}
+	wc := watcher.ResultChan()
+	for event := range wc {
+		csr, ok := event.Object.(*certificates.CertificateSigningRequest)
+		if !ok {
+			return nil, fmt.Errorf("unexpected event type during watch")
+		}
+		if event.Type == watch.Modified && csr.ObjectMeta.Name == csrObjectName {
+			for _, cond := range csr.Status.Conditions {
+				if cond.Type == certificates.CertificateApproved {
+					cert = csr.Status.Certificate
+					break
+				}
+			}
+		}
+		if len(cert) > 0 {
+			break // approved
+		}
+	}
+
+	keyBytes, err := x509.MarshalECPrivateKey(key.(*ecdsa.PrivateKey))
```

Comment:
Since the code is tightly coupled to ECDSA keys, I feel like having genPrivateKey return an *ecdsa.PrivateKey would be cleaner and obviate the need for this type assertion.

If you want to keep the code generic to key type, consider using MarshalPKCS8PrivateKey (along with the "PRIVATE KEY" PEM type) since both RSA and ECDSA private keys are supported.

Reply:
changed to use MarshalPKCS8PrivateKey  and "PRIVATE KEY"